### ISSUE TITLE
Slow relations initialization

### ISSUE THREAD

**Issue by 9too** [begriffs/postgrest#924]
# Context
PostgREST version : 0.4.2.0
PostgreSQL : 9.6.3
OS : Gentoo & RHEL 6.9

The source of the problem is unclear and I cannot provide a simple reproducible case.
( I provided @ruslantalpa with a _minimal_ database dump )

# Observation
On a particular database, requesting resource embedding for the first time causes a **significant** computation over a long period of time. On the sample provided, it takes ~30sec on my box and PostgREST was using ~40% of CPU load the whole time. Although that was after a cleanup, before I simply didn't wait long enough (> 10min). The query itself seems unimportant, as asking for an invalid relation triggers this problem. Hence, the exposed schema to test this simply contains one table such as
```sql
create table my_api.test ( col int );
```
And the problem happens when issuing a query like
```sh
curl 'http://localhost:3000/test?select=*,test(*)'
```

## Debugging
Everything involves the whole database and does _not_ involve the schema exposed to PostgREST.
- Trying to scope down the cause, it was found that :
   - Dropping one specific table completely removed the delay, albeit the table is nothing special at first glance
   - Dropping a couple views in a particular schema improves the delay gradually
- Requesting for a single entity while PostgREST is under this computation works without any delay.

**Comment by 9too** [comment]
With a bit more troubleshooting and a dose of luck, I have found the potential culprit.
The table which caused the problem to disapear contains 2 foreign keys which are exactly the same reference, they only have different names. Of course this is an error and shouldn't be there, but by dropping any of the two, the problem goes away.
Now I tried recreating a simple case with only 2 tables, but it does not cause any noticeable delay.
My guess is that it probably starts to mess up when there are a certain amount of references to be analyzed.

**Comment by ruslantalpa** [comment]
Interesting case, although i think there are valid cases for 2 Fks to the same table so this still is interesting to investigate.

**Comment by 9too** [comment]
After more testing it seems my assumption was incorrect, the problem is likely related but not exactly caused by identical FK duplicates.

I could find out by diffing a local database that I had fixed doing this with a preproduction version that was slightly altered from my version, where the problem still showed up after applying the above change.

In the diffing process, I found a FK that after being dropped, fixed the problem again. But this is a valid FK constraint that applies on a different column of the constrained table, but to the same reference.
The FK is on a character(8), that's the only particularity I can see.

Again I'm not able to reproduce with a minimal case ...

Another finding was that dropping all the RULES on a test improved the init time from 12s to 3s.
So it seems like there is a relation in the graph that causes everything else to slow down by an important factor.

At this point, Im going to start learning some haskell and dive into the code !

**Comment by 9too** [comment]
Here's a profiling of an unfinished initialization of a few seconds :  
```
COST CENTRE                         MODULE                SRC                                            %time %alloc

addManyToManyRelations.combinations PostgREST.DbStructure src/PostgREST/DbStructure.hs:256:5-63           86.4   99.4
addManyToManyRelations.links        PostgREST.DbStructure src/PostgREST/DbStructure.hs:253:5-117           6.0    0.0
addManyToManyRelations              PostgREST.DbStructure src/PostgREST/DbStructure.hs:(251,1)-(265,29)    5.8    0.0
```

Something odd to me, is that many-to-many takes so much time, but when I drop a FK that is a one-to-many, to problem disappears. After dropping that constraint, none of the sections above appear in the profiling.

**Comment by ruslantalpa** [comment]
This is a very good lead, at the root of combinations we have `subsequences`
```
The subsequences function returns the list of all subsequences of the argument.
subsequences "abc" == ["","a","b","ab","c","ac","bc","abc"]
```
which would be very inefficient on big arrays.
In this particular case, i think big arrays happen when there is a specific table with relations to multiple other tables, i.e. many of its columns are FKs to other tables.

I'll confirm that with your schema (you can also do that) then we'll think of another way to write that function which more optimal

**Comment by 9too** [comment]
I placed a Debug.Trace at the exit of subsequences, I will try to compare the results when that FK is present vs not.

**Comment by 9too** [comment]
Now, we found out that the definition of `addManyToManyRelations.combinations` can be improved for cases where alot of relations have to be computed.

I will try a some alternative definitions and compare them, then I'll open a PR.

**Comment by 9too** [comment]
After reading #796, it reminded me of a thought I also had;

If it would be possible to make `allRelations` (or create a new request if needs be) return only relations that are linked with the exposed schema ?

The problem is not so much the query itself, but rather how it can leverage the mapping done by `addManyToManyRelations` so that the minimum required relations are fetched.

If I compare some metrics I collected with changes in #926 between 2 databases : 

|| allRelations [count] | Startup time approx. (first query) [ms]
--- | --- | ---
|| 51 | 12
|| 458 | 331
**Ratio~** | 9 | 28

It simply shows that the time doesn't scale linearly with the amount of relations, and more importantly, independantly of the exposed schema. If at least it would scale directly with the relations exposed by the schema, it becomes easier to control the complexity and even to draw a parallel between what you control and the impact on the initialization.

This is just a thought, and I also consider this idea to be potentially overkill, and having a negligible benefit.

**Comment by ruslantalpa** [comment]
Negligible effect, for 1 request. This is overthi king it. It's good as it is now

**Comment by 9too** [comment]
Fixed in #926
